ding/reward_model/count_based_model.py

```diff
@@ -0,0 +1,433 @@
+    """
+    Overview:
+        The Pdeil reward model class
+    Interface:
+        ``estimate``, ``train``, ``load_expert_data``, ``collect_data``, ``clear_date``, \
+            ``__init__``, ``_train``, ``_batch_mn_pdf``
+    """
+    config = dict(
+        type='countbased',
+        counter_type='PixelCNN',
+        intrinsic_reward_type='add',
+        bonus_coeffient=0.1,
+        img_height=42,
+        img_width=42,
+        in_channels=1,
+        out_channels=1,
+        n_gated=2,
+        gated_channels=16,
+        head_channels=64,
+        q_level=256,
+    )
+
+    def __init__(
+        self,
+        cfg: dict,
+        device,
+        tb_logger: 'SummaryWriter'
+    ) -> None:  # noqa
+        """
+        Overview:
+            Initialize ``self.`` See ``help(type(self))`` for accurate signature.
+            Some rules in naming the attributes of ``self.``:
+                - ``e_`` : expert values
+                - ``_sigma_`` : standard division values
+                - ``p_`` : current policy values
+                - ``_s_`` : states
+                - ``_a_`` : actions
+        Arguments:
+            - cfg (:obj:`Dict`): Training config
+            - device (:obj:`str`): Device usage, i.e. "cpu" or "cuda"
+            - tb_logger (:obj:`str`): Logger, defaultly set as 'SummaryWriter' for model summary
+        """
+        super(BaseRewardModel, self).__init__()
+        self.cfg: dict = cfg
+        self._counter_type = cfg.counter_type
+        self.device = device
+        self.tb_logger = tb_logger
+        assert self._counter_type in ['SimHash', 'AutoEncoder', 'PixelCNN']
+        if self._counter_type == 'PixelCNN':
+            print(cfg)
+            self._counter = GatedPixelCNN(
+                in_channels=cfg.in_channels,
+                out_channels=cfg.out_channels,
+                n_gated=cfg.n_gated,
+                gated_channels=cfg.gated_channels,
+                head_channels=cfg.head_channels,
+                q_level=cfg.q_level,
+            )
+            self._counter.to(self.device)
+            self.intrinsic_reward_type = cfg.intrinsic_reward_type
+            self.index_range = np.arange(cfg.img_height * cfg.img_width)
+            assert self.intrinsic_reward_type in ['add', 'new', 'assign']
+            self.opt = optim.RMSprop(
+                self._counter.parameters(),
+                momentum=0.9,
+                weight_decay=0.95,
+                eps=1e-4,
+            )
+
+    def _train(self, train_data, t):
+        '''
+        Overview:
+            Using input data to train the network while estimating intrintic reward.
+        Arguments:
+            train_data (:obj:'torch.Tensor'): Observation with shape [1, 1, 42, 42].
+        '''
+        flattened_logits, target_pixel_loss, _ = self._counter(train_data)
+
+        # flattened_logits shape: [BHWC, D]; target_pixel_loss shape: [BHWC]
+        loss = nn.CrossEntropyLoss(reduction='none')(
```

Comment:
You should check the image input range of this reward model is in [0, 1], add its target should multiply 7(with pixel values quantized to 8 bins in original paper) and then transform into long type